Add more and better Rust examples

scuellar opened the pull request:
Following our Mattermost discussion about Rust, I've added better examples that more clearly showcase the differences between Rust and CN ownership/mutability models.

bcpierce00 commented:
Great!

Any plans to create some text that might be added to the tutorial?


On Tue, May 7, 2024 at 12:52 PM Santiago Cuellar ***@***.***>
wrote:

> Following our Mattermost discussion about Rust, I've added better examples
> that more clearly showcase the differences between Rust and CN
> ownership/mutability models.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/rems-project/cn-tutorial/pull/10
> <https://urldefense.com/v3/__https://github.com/rems-project/cn-tutorial/pull/10__;!!IBzWLUs!UUS6N6OxCBJtsfe1In5oslcv5UdP9FVeSmV3p-MhrB9r0hYYKWCrHQYFKll-j23ZVrAtD-VM2OyRlbghfEcY-DTrQwQE$>
> Commit Summary
>
>    - 57c0c30
>    <https://urldefense.com/v3/__https://github.com/rems-project/cn-tutorial/pull/10/commits/57c0c3048839d849dcc4b64de91314a5726dc3aa__;!!IBzWLUs!UUS6N6OxCBJtsfe1In5oslcv5UdP9FVeSmV3p-MhrB9r0hYYKWCrHQYFKll-j23ZVrAtD-VM2OyRlbghfEcY-L5cCRy2$>
>    Add examples qith clear comparisons.
>
> File Changes
>
> (7 files
> <https://urldefense.com/v3/__https://github.com/rems-project/cn-tutorial/pull/10/files__;!!IBzWLUs!UUS6N6OxCBJtsfe1In5oslcv5UdP9FVeSmV3p-MhrB9r0hYYKWCrHQYFKll-j23ZVrAtD-VM2OyRlbghfEcY-Plvkci5$>
> )
>
>    - *A*
>    src/example-archive/Rust/broken/error-crash/00008-pointer-compare.c
>    <https://urldefense.com/v3/__https://github.com/rems-project/cn-tutorial/pull/10/files*diff-f36188e799854d402c915f60f08c030146589574fc9f2cfeb39f279733a3d3a3__;Iw!!IBzWLUs!UUS6N6OxCBJtsfe1In5oslcv5UdP9FVeSmV3p-MhrB9r0hYYKWCrHQYFKll-j23ZVrAtD-VM2OyRlbghfEcY-J7w5Gls$>
>    (52)
>    - *A* src/example-archive/Rust/broken/error-crash/00012-move.c
>    <https://urldefense.com/v3/__https://github.com/rems-project/cn-tutorial/pull/10/files*diff-8011e9f3df1be67f5a546144be009dde018953977104363f366f4ab4bdda01e2__;Iw!!IBzWLUs!UUS6N6OxCBJtsfe1In5oslcv5UdP9FVeSmV3p-MhrB9r0hYYKWCrHQYFKll-j23ZVrAtD-VM2OyRlbghfEcY-IYUx1Wa$>
>    (79)
>    - *A* src/example-archive/Rust/broken/error-crash/00013-slices.c
>    <https://urldefense.com/v3/__https://github.com/rems-project/cn-tutorial/pull/10/files*diff-105a961608127ff0ba1c4d5d9a84d55045aa18df26ade87031da73cf1936939d__;Iw!!IBzWLUs!UUS6N6OxCBJtsfe1In5oslcv5UdP9FVeSmV3p-MhrB9r0hYYKWCrHQYFKll-j23ZVrAtD-VM2OyRlbghfEcY-KewIIDG$>
>    (87)
>    - *A* src/example-archive/Rust/broken/error-crash/00014-exception.c
>    <https://urldefense.com/v3/__https://github.com/rems-project/cn-tutorial/pull/10/files*diff-116a1c67d16ba13a348b43401c11582be5c3bc3fc05f7511e5097ff8fa1693fc__;Iw!!IBzWLUs!UUS6N6OxCBJtsfe1In5oslcv5UdP9FVeSmV3p-MhrB9r0hYYKWCrHQYFKll-j23ZVrAtD-VM2OyRlbghfEcY-CgriZco$>
>    (16)
>    - *A* src/example-archive/Rust/working/00009-read_write.c
>    <https://urldefense.com/v3/__https://github.com/rems-project/cn-tutorial/pull/10/files*diff-6aac0563f2002e29385d844f81c6218ea389fa0fd6ed35befb9e968521d53706__;Iw!!IBzWLUs!UUS6N6OxCBJtsfe1In5oslcv5UdP9FVeSmV3p-MhrB9r0hYYKWCrHQYFKll-j23ZVrAtD-VM2OyRlbghfEcY-GR1ACj8$>
>    (78)
>    - *A* src/example-archive/Rust/working/00010-references_call.c
>    <https://urldefense.com/v3/__https://github.com/rems-project/cn-tutorial/pull/10/files*diff-3bd86f3bfcfb64b5072aa7b63320d3df70dccc4a35acae297249299462f77b25__;Iw!!IBzWLUs!UUS6N6OxCBJtsfe1In5oslcv5UdP9FVeSmV3p-MhrB9r0hYYKWCrHQYFKll-j23ZVrAtD-VM2OyRlbghfEcY-MSufHRs$>
>    (89)
>    - *A* src/example-archive/Rust/working/00011-references.c
>    <https://urldefense.com/v3/__https://github.com/rems-project/cn-tutorial/pull/10/files*diff-9c8a405590c7ab28664d776e6a286e6fdd38ee60e1fbcfd966ab82bb7620f040__;Iw!!IBzWLUs!UUS6N6OxCBJtsfe1In5oslcv5UdP9FVeSmV3p-MhrB9r0hYYKWCrHQYFKll-j23ZVrAtD-VM2OyRlbghfEcY-PpD7vKz$>
>    (83)
>
> Patch Links:
>
>    - https://github.com/rems-project/cn-tutorial/pull/10.patch
>    <https://urldefense.com/v3/__https://github.com/rems-project/cn-tutorial/pull/10.patch__;!!IBzWLUs!UUS6N6OxCBJtsfe1In5oslcv5UdP9FVeSmV3p-MhrB9r0hYYKWCrHQYFKll-j23ZVrAtD-VM2OyRlbghfEcY-IcL3ArL$>
>    - https://github.com/rems-project/cn-tutorial/pull/10.diff
>    <https://urldefense.com/v3/__https://github.com/rems-project/cn-tutorial/pull/10.diff__;!!IBzWLUs!UUS6N6OxCBJtsfe1In5oslcv5UdP9FVeSmV3p-MhrB9r0hYYKWCrHQYFKll-j23ZVrAtD-VM2OyRlbghfEcY-MnM1PA8$>
>
> —
> Reply to this email directly, view it on GitHub
> <https://urldefense.com/v3/__https://github.com/rems-project/cn-tutorial/pull/10__;!!IBzWLUs!UUS6N6OxCBJtsfe1In5oslcv5UdP9FVeSmV3p-MhrB9r0hYYKWCrHQYFKll-j23ZVrAtD-VM2OyRlbghfEcY-DTrQwQE$>,
> or unsubscribe
> <https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/ABVQQCY2LMR2MKYRQN2XMBDZBEBDLAVCNFSM6AAAAABHLL765SVHI2DSMVQWIX3LMV43ASLTON2WKOZSGI4DGOBRG42DKMY__;!!IBzWLUs!UUS6N6OxCBJtsfe1In5oslcv5UdP9FVeSmV3p-MhrB9r0hYYKWCrHQYFKll-j23ZVrAtD-VM2OyRlbghfEcY-PZizvWc$>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

yav commented on src/example-archive/Rust/broken/error-crash/00008-pointer-compare.c:
```diff
@@ -0,0 +1,52 @@
+    println!("The number is: {}", r); // Safe to use 'r' here
+}
+
+fn borrow<'a>(input: &'a i32) -> &'a i32 {
```
It might be nice to make this be the same as the C code (i.e., write `compare` in Rust)

yav commented on src/example-archive/Rust/broken/error-crash/00012-move.c:
```diff
@@ -0,0 +1,79 @@
+}
+*/
+
+
```
I know this is not the point here, but would this be better with `Block` rather than `Owned`, as allocating and deallocating a string shouldn't really require it to be initialized?

scuellar replied:
Yes. I was trying to mimic what Rust does and reduce complexity. The comment for the function explains that it creates and initializes the string, but perhaps I should change the name as well to capture that.

yav commented on src/example-archive/Rust/broken/error-crash/00013-slices.c:
We are checking for 0, not space (the comment is wrong)

yav commented on src/example-archive/Rust/broken/error-crash/00013-slices.c:
Comment seems wrong, `s1` is returned?

yav commented on src/example-archive/Rust/broken/error-crash/00013-slices.c:
I wonder if it is better to do this example with a slice containing elements of a type that is not `Copy`, as it would make it more explicit what's happening to the ownership.   In particular, you wouldn't use the pattern in the `for` loop, so you'd have `for item in p`, and then at the use site you'd have to do a clone (i.e., `item.clone()`).

scuellar replied:
Could you say more about this, I'm not sure I understand.

yav commented on src/example-archive/Rust/working/00010-references_call.c:
```diff
@@ -0,0 +1,89 @@
+    let mut x:u32 = 42;
+    let s2 = &mut x; 
+    let s3:&mut u32 = &mut x; // Fails: can't have two mutable refs.
+    
```
I don't think you want the `*` here, we are just passing the reference?

yav commented on src/example-archive/Rust/working/00010-references_call.c:
```diff
@@ -0,0 +1,89 @@
+  unsigned int  s = 42;
+  unsigned int* s1 = &s; // Mutable reference to s
+  unsigned int* s2 = &s; // Another mutable reference to s
+
```
Comments are wrong, the references are `s1` and `s2`

scuellar commented:
@yav thank you for proof-checking the examples. I've addressed your comments except for the one I commented above. 

We could work together to improve that example if you are interested.

scuellar commented:
> Great! Any plans to create some text that might be added to the tutorial?

I don't know how @bcpierce00  and @cp526 want to deal with changes to the text. The new examples have some text explaining the differences I could gather that as tutorial text if that helps.

bcpierce00 commented:
In general, Christopher prefers (and I am fine with) just the two of us
having our hands in the main version of the actual document, except for
easy fixes.  So perhaps the easiest way to handle this sort of thing is as
a PR -- i.e., go ahead and add a new section to the tutorial but let one of
us review and polish before integrating it?


On Tue, May 7, 2024 at 3:23 PM Santiago Cuellar ***@***.***>
wrote:

> Great! Any plans to create some text that might be added to the tutorial?
>
> I don't know how @bcpierce00
> <https://urldefense.com/v3/__https://github.com/bcpierce00__;!!IBzWLUs!SQC16oayjNyPFHTnt-j-z7GuP4O4OeXiv6gXtwz01TRjoqEm6U8JGpfyzQq_ZttDxkGDtcomJsgQlsZvdHiTs89aYjn4$>
> and @cp526
> <https://urldefense.com/v3/__https://github.com/cp526__;!!IBzWLUs!SQC16oayjNyPFHTnt-j-z7GuP4O4OeXiv6gXtwz01TRjoqEm6U8JGpfyzQq_ZttDxkGDtcomJsgQlsZvdHiTs0cCfmUD$>
> want to deal with changes to the text. The new examples have some text
> explaining the differences I could gather that as tutorial text if that
> helps.
>
> —
> Reply to this email directly, view it on GitHub
> <https://urldefense.com/v3/__https://github.com/rems-project/cn-tutorial/pull/10*issuecomment-2099141922__;Iw!!IBzWLUs!SQC16oayjNyPFHTnt-j-z7GuP4O4OeXiv6gXtwz01TRjoqEm6U8JGpfyzQq_ZttDxkGDtcomJsgQlsZvdHiTs-fEJaqJ$>,
> or unsubscribe
> <https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/ABVQQCYRUOJYMJ4E76YJZGDZBESZJAVCNFSM6AAAAABHLL765SVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAOJZGE2DCOJSGI__;!!IBzWLUs!SQC16oayjNyPFHTnt-j-z7GuP4O4OeXiv6gXtwz01TRjoqEm6U8JGpfyzQq_ZttDxkGDtcomJsgQlsZvdHiTsySi9AS9$>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

septract commented:
@scuellar is this still in progress or should we move to merge it?

bcpierce00 commented:
What's the status of this stuff?  If it's going in the tutorial, then getting it ready in the next few days would be great -- I'd like to do another round of work on the tutorial and get it to some kind of "new level of stable."

scuellar commented:
I was trying to find some time to turn this into a tutorial write up, but it seems like that won't be happening soon and I need to focus more in the CN engineering. I'll clean up the examples and mark this ready for merge. 

@bcpierce00  If anyone wants to port the examples into the tutorial I'm happy to serve as a resource. Otherwise I can put this into my stack and get to it when I get more time.

bcpierce00 commented:
I'm afraid I'm not a Rust programmer, so I'm not the one to do this bit of writing.  What I was imagining was a collection of matched pairs -- here's a Rust program and here's how you achieve the same effect in CN -- with a bit of commentary about what's going on in both the Rust and the CN versions.  I do think this would be extremely useful -- both for helping onboard Rust programmers and for broader advertising purposes -- so I hope your cleanup process can at least include leaving enough notes / bullet points that you can recover your thought process later.
